### PR TITLE
Ag 12568/annotation options toolbar drag

### PR DESCRIPTION
**Pull request by manapeirov**
PR for https://ag-grid.atlassian.net/browse/AG-12568

### CODE REVIEW

File: packages/ag-charts-community/src/chart/toolbar/toolbar.ts [hunk 1]
```diff
@@ -676,6 +727,12 @@ export class Toolbar extends BaseModuleInstance implements ModuleInstance {
         button.onclick = makeAccessibleClickListener(button, (event) =>
             this.onButtonPress(event, button, group, options.id, options.value)
         );
+
+        if (options.value === 'drag') {
+            button.onmousedown = makeAccessibleClickListener(button, (event) => this.onDragStart(event, group));
```
Comment (lsjroberts) [inline packages/ag-charts-community/src/chart/toolbar/toolbar.ts]:
```suggestion
            button.addEventListener('mousedown', makeAccessibleClickListener(button, (event) => this.onDragStart(event, group)));
```

File: packages/ag-charts-community/src/chart/toolbar/toolbar.ts [hunk 2]
```diff
@@ -770,6 +827,46 @@ export class Toolbar extends BaseModuleInstance implements ModuleInstance {
...
+                Number(element.style.getPropertyValue('left').replace('px', '')),
+                Number(element.style.getPropertyValue('top').replace('px', ''))
+            ),
+            detached: true,
+        };
+
+        const onDrag = (e: MouseEvent) => this.onDrag(e, group);
+        const window = getWindow();
+        window.addEventListener('mousemove', onDrag);
+        window.addEventListener('mouseup', () => window.removeEventListener('mousemove', onDrag), {
```
Comment (lsjroberts) [inline packages/ag-charts-community/src/chart/toolbar/toolbar.ts:840]: I was having some issues with this with the dialog since some events are not propagated up beyond our container element. I suggest using the `domManager` instead of `window` as that contains all we care about and will always receive the events.

```suggestion
        const { ctx: { domManager } } = this;
        domManager.addEventListener('mousemove', onDrag);
        domManager.addEventListener('mouseup', () => domManager.removeEventListener('mousemove', onDrag), {
```

File: packages/ag-charts-community/src/chart/toolbar/toolbarTypes.ts [hunk 3]
```diff
@@ -45,3 +45,9 @@ export interface ToolbarButton extends ToolbarButtonConfig {
     role?: 'button' | 'switch';
     checkedOverrides?: ToolbarButtonConfig;
 }
+
+export interface Anchor {
```
Comment (lsjroberts) [inline packages/ag-charts-community/src/chart/toolbar/toolbarTypes.ts]: To be more specific:

```suggestion
export interface ToolbarAnchor {
```

File: packages/ag-charts-community/src/chart/toolbar/toolbar.ts [hunk 4]
```diff
@@ -285,16 +302,16 @@ export class Toolbar extends BaseModuleInstance implements ModuleInstance {
...
+
+        const { group, anchor, index } = event;
+
+        if (this.index == index && this.dragState.detached) {
+            return;
+        }
 
-        const { group, anchor } = event;
+        this.index = index;
+        this.dragState.detached = false;
```
Comment (lsjroberts) [inline packages/ag-charts-community/src/chart/toolbar/toolbar.ts]: This `index` is very specific to annotations and a bit of a random number to save here in toolbar. It's a bit leaky to have it here.

I _think_ you could achieve the same behaviour of reattaching the floating toolbar when you select a new annotation by resetting `detached` in the `Toolbar.toggleVisibilities()` function, something like:

```diff
for (const position of TOOLBAR_POSITIONS) {
    const visible = this.enabled && Array.from(this.positions[position].values()).some(isGroupVisible);
    this.elements[position].classList.toggle(styles.modifiers.hidden, !visible);
+   if (position === ToolbarPosition.Floating && !visible) {
+       this.dragState.detached = false;
+   }
}
```

As when you select a new annotation it will deselect the previous one first, hide the overlays, toggling the visibility of the floating toolbar to false and reattaching it here, then show the new version of it.

File: packages/ag-charts-community/src/chart/toolbar/toolbar.ts [hunk 5]
```diff
@@ -512,6 +557,12 @@ export class Toolbar extends BaseModuleInstance implements ModuleInstance {
                 this.setButtonActive(button, active);
             }
         }
+
+        if (this[group].draggable) {
+            this.ctx.toolbarManager.toggleButton(group, 'drag', {
+                enabled,
+            });
+        }
```
Comment (lsjroberts) [inline packages/ag-charts-community/src/chart/toolbar/toolbar.ts]: Calling from `toolbar` out to the `toolbarManager` to trigger a listener inside `toolbar` is a bit circuitous.

What do you need from the `onButtonToggled()` listener to happen? Other than the `setButtonActive()` that's already called a few lines above?

File: packages/ag-charts-community/src/chart/toolbar/toolbar.ts [hunk 6]
```diff
@@ -622,7 +673,7 @@ export class Toolbar extends BaseModuleInstance implements ModuleInstance {
             if (this[group] == null) continue;
             const groupVisible = isGroupVisible(group);
             for (const button of this.groupButtons[group]) {
-                const buttonVisible = groupVisible && this[group].buttonConfigurations().some(isButtonVisible(button));
+                const buttonVisible = groupVisible && isButtonVisible(button);
```
Comment (lsjroberts) [inline packages/ag-charts-community/src/chart/toolbar/toolbar.ts]: Did you mean to commit this?

You'll need to revert this change, the `isButtonVisible(button)` call returns a function so will always be truthy. And we need the `buttonConfigurations()` bit to fix issues with locale changes resetting buttons I believe (see https://github.com/ag-grid/ag-charts/commit/d4e833f3c766837ce7898e83eea3603aeda9d0f5).

**Review by lsjroberts — CHANGES_REQUESTED** [review]

File: packages/ag-charts-community/src/chart/toolbar/toolbar.ts (hunk 2 above)
Comment (lsjroberts) [reply packages/ag-charts-community/src/chart/toolbar/toolbar.ts:840]: Though, may not be needed now - https://github.com/ag-grid/ag-charts/pull/2347/commits/c17cc10d94298fc5b1e71c20dbd96fb5319f7017

Comment (manapeirov) [reply packages/ag-charts-community/src/chart/toolbar/toolbar.ts:840]: The issue I found with using domManager is that dragging will continue if you release the mouse (mouseup) outside of the canvas container, I did window so it behaves like the navigator drag

File: packages/ag-charts-community/src/chart/toolbar/toolbar.ts (hunk 5 above)
Comment (manapeirov) [reply packages/ag-charts-community/src/chart/toolbar/toolbar.ts]: I think I left that in there by accident :)

File: packages/ag-charts-community/src/chart/toolbar/toolbar.ts (hunk 4 above)
Comment (manapeirov) [reply packages/ag-charts-community/src/chart/toolbar/toolbar.ts]: Discussed offline and decided to fix in refactor of toobar later

**Review by lsjroberts — APPROVED** [review]
LGTM, excellent stuff!
